File: apps/user/src/dao/ImageDao.ts
'use strict'

import {ObjectList} from "aws-sdk/clients/s3";

const AWS = require('aws-sdk')
import {S3} from "aws-sdk";

class ImageDao {
    s3: S3
    bucket: string
    constructor() {
        this.s3 = new AWS.S3();
        // @ts-ignore
        this.bucket = process.env.IMAGE_BUCKET
    }

    async uploadImage(userId: string, decodedImage: Buffer, fileName: string, contentType: string): Promise<string> {
        const params = {
            "Body": decodedImage,
            "Bucket": this.bucket,
            "Key": `users/${userId}/images/${fileName}`,
            "ContentType": contentType,
            'ACL': 'public-read'
        }

        const res = await this.s3.upload(params).promise();
        return res.Location
    }

    async removeOldImages(userId: string, excludedFile: string) {
        const objectList = await this.s3.listObjectsV2({
            Bucket: this.bucket,
            Delimiter: '/',
            Prefix: `users/${userId}/images/`
        }).promise()

        const images: ObjectList = objectList.Contents ? objectList.Contents : []
        for (const image of images) {
            if (!image.Key?.includes(excludedFile)) {
                await this.s3.deleteObject({
                    Bucket: this.bucket,
                    Key: image.Key!
                }).promise()
            }
        }
    }

    async getImage(key: string): Promise<S3.GetObjectOutput> {
        const params = {
            "Bucket": this.bucket,
            "Key": key
        }

        return await this.s3.getObject(params).promise();
    }
}

export {ImageDao}

File: apps/user/src/dao/image_dao.py
import os

import boto3
from types_boto3_s3 import S3Client


class ImageDao:
    def __init__(self):
        self.s3: S3Client = boto3.client("s3")
        self.bucket = os.environ["IMAGE_BUCKET"]

    def upload_image(self, user_id: str, image, file_name: str, content_type: str) -> str:
        key = f"users/{user_id}/images/{file_name}"
        self.s3.upload_fileobj(
            Fileobj=image,
            Bucket=self.bucket,
            Key=key,
            ExtraArgs={'ContentType': content_type, 'ACL': 'public-read'}
        )

        return f"https://{self.bucket}.s3.amazonaws.com/{key}"

    def remove_old_images(self, user_id: str, excluded_key: str):
        objects = self.s3.list_objects_v2(
            Bucket=self.bucket,
            Delimiter='/',
            Prefix=f"users/{user_id}/images/"
        )

        keys_to_remove = [obj.get("Key") for obj in objects.get("Contents", []) if excluded_key not in obj.get("Key")]
        for key in keys_to_remove:
            self.s3.delete_object(Bucket=self.bucket, Key=key)
